File: leetcode/string/14.py
from typing import List


class Solution:

    def longestCommonPrefix(self, strs: List[str]) -> str:
        '''最长公共前缀
        
        @Note:
            纵向比较
        '''
        if len(strs) == 0:
            return ''
        for j in range(len(strs[0])):
            for i in range(1, len(strs)):
                if len(strs[i]) < j + 1 or strs[i][j] != strs[0][j]:
                    return strs[0][:j]
        return strs[0]  #注意这,如果前面条件都满足,那么返回strs[0]


if __name__ == '__main__':
    string = input().strip()
    solution = Solution()
    print(solution.longestCommonPrefix(string))

File: leetcode/string/14.cpp
#include <algorithm>
#include <iostream>
#include <vector>

class Solution {
 public:
  std::string longest_common_prefix(std::vector<std::string>& strs) {
    auto min_iter = std::min_element(strs.begin(), strs.end(), [&](const std::string& lhs, const std::string& rhs) {
      return lhs.size() < rhs.size();
    });

    for (int i = 0; i < static_cast<int>((*min_iter).size()); ++i) {
      char s = strs[0][i];
      for (size_t j = 1; j < strs.size(); ++j) {
        if (strs[j][i] != s) {
          return (i - 1 < 0) ? "" : strs[j].substr(0, i);
        }
      }
    }
    return *min_iter;
  }
};

int main() {
  std::vector<std::string> vec;
  vec.emplace_back("floor");
  vec.emplace_back("flower");
  vec.emplace_back("flew");

  Solution solution = Solution();
  std::string pre = solution.longest_common_prefix(vec);

  std::cout << "pre: " << pre << std::endl;
}